src/test/java/com/google/cloud/spanner/pgadapter/python/PythonTransactionTests.java

```diff
@@ -0,0 +1,1514 @@
+    String actualOutput = executeTransactions(pgServer.getLocalPort(), statements);
+
+    assertEquals(expectedOutput, actualOutput);
+
+    assertEquals(0, mockSpanner.countRequestsOfType(CommitRequest.class));
+    assertEquals(0, mockSpanner.countRequestsOfType(RollbackRequest.class));
+    assertEquals(0, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+  }
+
+  @Test
+  public void testDeferrableSessionInTransactions() throws IOException, InterruptedException {
+    List<String> statements = new ArrayList<>();
+    String sql = "Select * from some_table";
+
+    // It will throw an error because deferrable is not supported
+    statements.add("transaction");
+    statements.add("set session deferrable True");
+
+    // statement won't be executed because the above setting will throw error
+    statements.add("query");
+    statements.add(sql);
+
+    mockSpanner.putStatementResult(
+        StatementResult.query(Statement.of(sql), createResultSet(1, "abcd")));
+
+    String expectedOutput = "the 'deferrable' setting is only available from PostgreSQL 9.1\n";
+    String actualOutput = executeTransactions(pgServer.getLocalPort(), statements);
+
+    assertEquals(expectedOutput, actualOutput);
+
+    assertEquals(0, mockSpanner.countRequestsOfType(CommitRequest.class));
+    assertEquals(0, mockSpanner.countRequestsOfType(RollbackRequest.class));
+    assertEquals(0, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+  }
+
+  @Test
+  public void testReadOnlyInTransactions() throws IOException, InterruptedException {
+    // tests readonly settings using connection.readonly variable
+    List<String> statements = new ArrayList<>();
+    String sql1 = "Select * from some_table";
+    String sql2 = "insert into some_table(col1, col2) values(value1, value2)";
+    String sql3 = "Select * from some_table3";
+    String sql4 = "insert into some_table3(col1, col2) values(value1, value2)";
+
+    // query will be executed as expected
+    statements.add("query");
+    statements.add(sql1);
+
+    // update will be executed as expected
+    statements.add("update");
+    statements.add(sql2);
+
+    // update will be executed as expected
+    statements.add("update");
+    statements.add(sql4);
+
+    // 1 commit request will be sent
+    statements.add("transaction");
+    statements.add("commit");
+
+    statements.add("transaction");
+    statements.add("set readonly True");
+
+    // query will be executed as expected
+    statements.add("query");
+    statements.add(sql1);
+
+    // query will be executed as expected
+    statements.add("query");
+    statements.add(sql3);
+
+    // commit request will not be sent to the Cloud Spanner
+    // because Cloud Spanner does not require read-only transactions
+    // to be committed or rolled back
+    statements.add("transaction");
+    statements.add("commit");
+
+    // update will throw an error because the readonly is activated
```

Comment:
According to the Psycopg2 documentation, it will set the whole session as readonly. It doesn't send any statement like `set default_transaction_read_only = true` or something but all the transactions started after `set readonly true` will be in the read only mode. You can take a look at the following Wire Messages received be the PGAdapter to check more on this.

> Received Message: (Q) Query, with Payload: {Length: 10, SQL: BEGIN}
> Received Message: (Q) Query, with Payload: {Length: 29, SQL: Select * from some_table}
> Received Message: (Q) Query, with Payload: {Length: 62, SQL: insert into some_table(col1, col2) values(value1, value2)}
> Received Message: (Q) Query, with Payload: {Length: 63, SQL: insert into some_table3(col1, col2) values(value1, value2)}
> Received Message: (Q) Query, with Payload: {Length: 11, SQL: COMMIT}
> Received Message: (Q) Query, with Payload: {Length: 36, SQL: BEGIN;SET TRANSACTION READ ONLY}
> Received Message: (Q) Query, with Payload: {Length: 29, SQL: Select * from some_table}
> Received Message: (Q) Query, with Payload: {Length: 30, SQL: Select * from some_table3}
> Received Message: (Q) Query, with Payload: {Length: 11, SQL: COMMIT}
> Received Message: (Q) Query, with Payload: {Length: 36, SQL: BEGIN;SET TRANSACTION READ ONLY}
> Received Message: (Q) Query, with Payload: {Length: 62, SQL: insert into some_table(col1, col2) values(value1, value2)}